source/base/NexusApp.cc

```diff
@@ -99,32 +99,32 @@ NexusApp::NexusApp(G4String init_macro): G4RunManager(), gen_name_(""),
   if (gen_name_ == "") {
     G4Exception("[NexusApp]", "NexusApp()", FatalException, "A generator must be specified.");
   }
-  pg->SetGenerator(std::unique_ptr<G4VPrimaryGenerator>(ObjFactory<G4VPrimaryGenerator>::Instance().CreateObject(gen_name_)));
+  pg->SetGenerator(ObjFactory<G4VPrimaryGenerator>::Instance().CreateObject(gen_name_));
   this->SetUserAction(pg);
 
   if (pm_name_ == "") {
     G4Exception("[NexusApp]", "NexusApp()", FatalException, "A persistency manager must be specified.");
   }
-  PersistencyManagerBase* pm = ObjFactory<PersistencyManagerBase>::Instance().CreateObject(pm_name_);
+  std::unique_ptr<PersistencyManagerBase> pm = ObjFactory<PersistencyManagerBase>::Instance().CreateObject(pm_name_);
   pm->SetMacros(init_macro, macros_, delayed_);
 
  // PersistencyManager::Initialize(init_macro, macros_, delayed_);
 
   // Set the user action instances, if any, in the run manager
   if (runact_name_ != "")
-    this->SetUserAction(ObjFactory<G4UserRunAction>::Instance().CreateObject(runact_name_));
+    this->SetUserAction(ObjFactory<G4UserRunAction>::Instance().CreateObject(runact_name_).get());
```

Comment:
This looks wrong.

The 'unique' in `unique_ptr` means that the object (the smart pointer, that is) can only exist in one place at a time: it can be *moved* from place to place, but it cannot be *copied*.

If you pass it to a function that takes a `unique_ptr<T>` you must move it from the caller into the function. If you return one from a function, it must be moved out of the function to the caller. This is what is happening here: 

1. The `unique_ptr` is moved (implicitly, because it's an rvalue) out of `CreateObject`. `CreateObject` no longer owns the  resource that the `unique_ptr` points to.
2. You use `.get()` to borrow (in other words, get a non-owning, observing handle: a raw pointer to) the resource, and lend it to `SetUserAction`.
3. The `unique_ptr` is dropped on the floor, so it gets destructed, so it destructs the resource it points to ... the same resource you just lent to `SetUserAction`.

Consequently, `SetUserAction` has been given a pointer which is guaranteed to be invalidated immediately.

By not moving the `unique_ptr` anywhere, not giving it to anyone to keep, but dropping it on the floor, you are saying 'nobody wants this any more', at which point the resource it owns gets deleted.

Who is responsible for the lifetime of the thing the factory just made?

+ If it's Geant4 via `SetUserAction`, (and Geant4 doesn't understand/use smart pointers (even though they were a crucial addition to the standard over a decade ago)) then use the `.release()` method of `unique_ptr` instead of `.get()` to say "I'm relinquishing ownership, it's your problem now".

+ If your code owns this thing (is responsible for deleting it), then make sure that you move the `unique_ptr` you got from the factory, into whatever it is that is supposed to own it.

Reply:
> Consequently, `SetUserAction` has been given a pointer which is guaranteed to be invalidated immediately.

Which is another example of nonsense that is simply not allowed by the Rust compiler.

Reply:
I've tried to implement what you said, let me know if that's the correct way.

Reply:
Yes, this looks much more plausible. Now:

+ `this->XXXact_` is the owner and is responsible for its lifetime,
+ `SetUserAction` gets to look at it and use it, but is not responsible for its lifetime.

But it begs the question: who was deleting these before? Or is this another instance of what this PR was trying to fix: these weren't being deleted at all?

Reply:
I suspect they weren't deleted (unless Geant4 did it itself).

Reply:
> unless Geant4 did it itself.

Eeeeek!

It's important to know whether Geant4 releases it or not! (And because G4 doesn't use smart pointers, you can't tell by just looking at the signature ... which is half of the point of these smart pointers.) 

+ If G4 *does* delete them, then what you have done is very wrong, and will result in double-deletion (G4 owns it, you also think you own it, and both of you delete it). In which case you should `.release()` into `SetUserAction`, and drop the `unique_ptr` on the floor. 
+ If G4 does *not* delete them, then your code looks correct.

Reply:
Interesting point. I found this:

https://geant4.kek.jp/lxr/source/run/src/G4RunManager.cc#L216

It seems that in the case of the actions Geant4 does delete them, right? I'm not sure I understand the comment: is it saying that we should not take over the ownership and set to zero the pointers in the destructor?

Reply:
Let's just take a moment to appreciate how much of our time is being wasted by this comment, and how much time we would save if G4 used smart pointers as the Standard Committee intended, rather than whatever ad-hoc scheme this is.

Now, what does it mean?

> [...] for all RunManager types if derived class [...]

[Here](https://gitlab.cern.ch/geant4/geant4/-/blob/master/source/run/include/G4RunManager.hh#L154) we see that `G4RunManager` has `virtual` methods (destructor, in this case) so it is meant to be subclassed. So 'all RunManager types' probably means 'all subclasses of `G4RunManager`'.

> [...] does the correct thing in the derived destructor that is: set to zero pointers of user initialization objects for which have no ownership

+ `delete`-ing a `nullptr` is guaranteed to be safe, essentially a null-op.
+ [`DeleteUserInitializations`](https://gitlab.cern.ch/geant4/geant4/-/blob/master/source/run/src/G4RunManager.cc#L244) `delete`s a bunch of member pointers.

So it seems to be saying that the onus is on the implementor of subclasses of `G4RunManager` to nuillify pointers it does not own, in the destructor, because this base class' destructor will `delete` anything that hasn' been nullified. Effectively, if it ain't `NULL` at destruction time, I'll assume that I own it and I will `delete` it.

Which tells us absolutely nothing about how the subclass of `G4RunManager` and its clients are supposed to decide who owns what. Maybe this decision process is documented somewhere? In general, it would seem that you have to go to the specific subclass of `G4RunManager` that you happen to be using and find out which of the things that you gave it, are nullified in its destructor (which might be done by a superclass' destructor). Have fun and good luck!

And now let's remind ourselves how this would look if G4 used smart pointers as intended:

+ If G4 takes ownership, it accepts the pointer as a `unique_ptr` and the compiler makes it clear to the client that ownership is being taken, because the client will either have to create a `unique_ptr` or `std::move` an existing `unique_ptr`, when calling the function, before the compiler is happy.

+ If G4 does not take ownership, it accepts the a bare pointer (or a reference) and the client knows that it keeps ownership.

The first case is always obvious because the compiler can enforce it. The second case, *crucially* requires G4 to be playing by the rules of smart pointers, which it isn't. In such cases the C++ compiler (unlike the Rust compiler) has absolutely no way of helping you. And that's the situation in which we find ourselves here. Yippeee!

Reply:
After discussing with @jmalbos we've come to the conclusion that, given that Geant4 does not use smart pointers, it's cumbersome and potentially error-prone to keep track of each pointer given to its classes. So, we think that it's better to use smart pointers only when nexus classes are the owners for all their lifetime. In this PR that means going back to commit 26de80f5.

Reply:
> given that Geant4 does not use smart pointers, [...] we think that it's better to use smart pointers only when nexus classes are the owners for all their lifetime.

Whether or not a nexus class should hold a smart pointer, is *completely independent* of whether or not Geant4 is smart pointer aware!

I get the impression that we haven't understood the rules of smart pointer usage. Given how useful and important (and in comparison to most other C++ features, **simple**) they are, it's crucial that we get this right. Let me have another go at explaining them, this time with a bit more detail, and explicitly covering legacy libraries. (I talk only about `unique_ptr`, but the rules are very similar for `shared_ptr`.) It may seem confusing and complicated at first, especially if there are legacy libraries to deal with, but, trust me, it really is one of the simpler and most useful things in C++.

1. Producers:

   + Legacy producers will return bare pointers. There is nothing you can do about this. But, if you don't *immediately* pass it to a legacy consumer, you should wrap it in a `unique_ptr` asap.
   + Modern producer: *Every* factory that you write in C++11 and beyond should return a `std::unique_ptr<T>`, without exception. (Yes, I did come up with a possible exception upthread. Let's ignore such obscure possibilities until we have thoroughly understood the fundamental rules of the game.)

2. Observers (things that use but do not own the `T` made by a producer):

   + It doesn't matter whether the observer be legacy or modern, in both cases it will take `T*` or `T&`. The owner of the `T` will use `std::unique_ptr::get()` to give the bare pointer or reference to the observer.

3. Consumers (this includes owners):

   + Legacy consumers will take (and in the case of owners, also keep) `T*` or `T&`. Whoever passes ownership to a legacy consumer, must use `std::unique_ptr::release()` to relinquish ownership to the consumer. The consumer will have to take care to `delete` the `T` at some stage, or pass ownership on to somebody else (but crucially, not both). Distinguishing legacy consumers from legacy observers, and ensuring that they `delete` every resource exactly once between them, is one of the major challenges of C++ programming.
   + Modern consumers will take (and in the case of owners, also keep) `std::unique_ptr<T>`. Whoever passes ownership to a modern consumer must use `std::move` to relinquish ownership to the consumer (though the move is implicit in the case of rvalues). The consumer won't have to `delete` the `T`, because it happens automatically when the `unique_ptr` is destructed, it will happen automatically regardless of whether ownership has been passed on or not; and it will happen exactly *once*, no more, no less. This is a big deal. It's a HUGE deal. Take the deal. Embrace the deal. It's one of the biggest and best deals there is in C++. Thank your lucky stars (and the standard committee) that the deal is now available.

Applying this to Nexus and Geant4:

+ Every factory you write in Nexus, should return a `unique_ptr<T>`
+ If the factory's product is owned by Geant4 and observed (used) by nexus:

  ```c++
  auto thing = thing_factory();
  nexus_observer(thing.get());
  geant4_owner(thing.release());
  ```
  NB: after `thing.release()`, `thing.get() == nullptr`

  ```c++
  class nexus_class_that_does_NOT_own_T {
    T*;
  };
  ```


+ If the factory's product is observed by Geant4 and owned by nexus:

  ```c++
  auto thing = thing_factory();
  geant4_observer(thing.get());
  nexus_owner(std::move(thing));
  ```
  NB: (generally speaking, not just in the case of `unique_ptr`) after `std::move(thing)`, `thing` is in an unusable state (unless you re-initialize it). `std::move()` effectively means: 'Dear compiler, I promise that `thing` will not be used again'. (In C++ it's up to the programmer to make sure this is the case; in Rust, the compiler makes sure that objects are not used after having been moved.)

  ```c++
  class nexus_class_that_DOES_own_T {
    std::unique_ptr<T>;
  };
  ```

> we think that it's better to use smart pointers only when nexus classes are the owners for all their lifetime.

While it is not uncommon for ownership of a resource to be transferred during its lifetime (well, it's guaranteed at least once: from factory to first owner), I suspect that this won't be happening much in nexus. Which gives me the impression that you have some, erm ... unorthodox ideas about what ownership is. (However, if nexus really is the temporary owner of some resource, and you think that it's better not to use smart pointers in such cases, well, I'm afraid you are simply worng. In fact, that would be an example of a situation where it's especially important to use smart pointers!)

I *suspect* that in most of these cases the fundamental question is: "This thing I just passed to Geant4, will Geant4 delete it?"

+ If the answer is yes:

  - Store it as `T*` in a nexus class.
  - `.get()` when giving it to the nexus class.
  - `.release()` when passing it to Geant4.
  
+ If the answer is no:

  - Store it as `std::unique_ptr<T>` in a nexus class.
  - `.get()` when passing it to Geant4.
  - `std::move()` when giving it to the nexus class.
  
NB: `.get()` must precede `.release()` or `std::move()`.

One more thing: *always* fully qualify `std::move`. That is, *never* `using std::move; [...] move()`.

> [...] it's cumbersome and potentially error-prone to keep track of each pointer given to its classes. [...]

Not sure what exactly is meant by this, but I can assure you that doing anything other than following the rules set out above is a lot more cumbersome and error-prone following the rules.

At the core of all of this is a very simple idea: 

+ If I have a `unique_ptr<T>` it means that I am the *sole owner* of this `T`. <sup><sub>If I wish, I can give my `T` away to someone else with `std::move()` (or `.release()`, if the new owner is legacy code).</sub></sup>
+ If I have a `T*` or `T&` it means that I am looking at someone else's `T`. (NB: in *legacy* code, it means absolutely nothing.)

Reply:
I think this is clear.

> [...] it's cumbersome and potentially error-prone to keep track of each pointer given to its classes. [...]
Not sure what exactly is meant by this, but I can assure you that doing anything other than following the rules set out above is a lot more cumbersome and error-prone following the rules.

What I meant is that it is not always straightforward to know the answer to

>  "This thing I just passed to Geant4, will Geant4 delete it?"

That's why we thought it was less error prone not to use smart pointers for pointers that are passed to Geant4.
Maybe one way to be sure that we don't delete pointers twice (once in nexus and a second time in Geant4) is to delete them always in nexus and then set them to `nullptr` afterwards, as the G4RunManager class seems to suggest?

Reply:
>> What I meant is that it is not always straightforward to know the answer to

> "This thing I just passed to Geant4, will Geant4 delete it?"

Of course! This difficulty is part and parcel of using legacy libraries!

Regardless of how difficult it is to answer this question, you *still* have to find the correct answer, not just in this case, but in all cases where you pass pointers. Using smart pointers can only make it easier to reason about these questions and write correct programs, it cannot make it more difficult. You cannot diminish the difficulty by not using smart pointers.

That is why you should migrate code that you own, away from this legacy style. Use smart pointers consistently in your code, and your clients will be able to answer this question trivially when reasoning about your code. In other words, don't inflict on your clients the same kind of pain that Geant4 is inflicting on you by not using smart pointers.

> That's why we thought it was less error prone not to use smart pointers for pointers that are passed to Geant4.

I trust that we agree that any pointer you pass to Geant4 (today) will *always* be a bare pointer. If you're suggesting that it will be less error prone to avoid using smart pointers altogether for anything that will eventually end up being passed to Geant4, then I *strongly* disagree.

> Maybe we can do as you suggest and one way to be sure that we don't delete pointers twice (once in nexus and a second time in Geant4) is to delete them always in nexus and then set them to nullptr afterwards, as the G4RunManager class seems to suggest?

In other words, you are suggesting to avoid the cost of answering the question "Will G4 delete this?" by hacking G4 in an ad-hoc way, tracking down every pointer you ever passed it, and ensuring that G4 *never* deletes any of them, so that you know that you have to delete them in your code? Please don't torture yourselves (and your users) like this! Madness lies this way.

The C++ standard committee provided you (pretty decent, as C++ goes) tools for ensuring that (dynamically allocated) resources get deleted exactly once: the C++11 standard smart pointers (`uniqe_ptr` and `shared_ptr`, not to be confused with the C++98 `auto_ptr`, which was *fundamentally broken*). Please don't avoid using them in favour of your own schemes that you make up on the spot! This cannot end well. 

You **have** to answer the crucial ownership question. There's no way around it. (Well, there is: write buggy code.) 

If you believe [this](https://gitlab.cern.ch/geant4/geant4/-/blob/master/examples/basic/B1/exampleB1.cc#L105-L107), then you should conclude that

+ user actions
+ physics lists
+ detector description

are owned by the run manager. If you need to store these in your own classes or pass them around your own functions, do so via bare pointers.

Way at the top of this PR, you added a deletion of a generator in a nexus class' destructor. It seems that G4 does not take ownership of these, and your nexus class does. Store these in your class via `unique_ptr<T>`, but pass them around via `T*`. Except when transferring them from the factory into the class: that should be done with `unique_ptr`.

For anything else, you'll have to find the correct answer to the ownership question, and act accordingly.

Even if you don't store it in your code at all and pass it to G4 as soon as it is made, you should probably still use smart pointers! It might be edifying to put yourselves in the shoes of a (smart pointer aware - and if you're not smart pointer aware in 2022, you shouldn't be programming in C++) programmer reading your code > 6 months in the future, and compare and contrast these alternatives 

1. ```c++
   auto t = my_factory();
   some_g4_instance.set_something(t);
   ```

2. ```c++
   auto t = my_factory();
   some_g4_instance.set_something(t.get());
   ```
   
3. ```c++
   auto t = my_factory();
   some_g4_instance.set_something(t.release());
   ```

+ What do these different samples tell you?
+ One of them should be ringing alarm bells, and have you asking to see the next few lines of the code! (Actually, *two* of them should be ringing alarm bells: one because of a very specific and obvious problem which might be solved on the next line; another because of the general paranoia any decent C++ programmer should have.)


Key takeaway:

+ In C++, you *must* know who will delete dynamically allocated resources (however difficult it may be to find out).
+ Smart pointers can only make this easier, not more difficult.